core/store/src/utils/mod.rs

```diff
@@ -324,6 +369,40 @@ pub fn remove_account(
         state_update.remove(TrieKey::AccessKey { account_id: account_id.clone(), public_key });
     }
 
+    // Removing gas keys
+    let lock = state_update.trie().lock_for_iter();
+    // Collect to hashset, as there can be multiple nonces per gas key.
+    let gas_public_keys = state_update
+        .locked_iter(&trie_key_parsers::get_raw_prefix_for_gas_keys(account_id), &lock)?
+        .map(|raw_key| {
+            trie_key_parsers::parse_public_key_from_gas_key_key(&raw_key?, account_id).map_err(
+                |_e| {
+                    StorageError::StorageInconsistentState(
+                        "Can't parse public key from raw key for GasKey".to_string(),
+                    )
+                },
+            )
+        })
+        .collect::<Result<HashSet<_>, _>>()?;
+    drop(lock);
+
+    for public_key in gas_public_keys {
+        let gas_key = get_gas_key(state_update.trie(), account_id, &public_key)?;
+        let Some(gas_key) = gas_key else {
+            tracing::warn!(
+                target: "store",
+                ?account_id,
+                ?public_key,
+                "gas key missing when removing account"
+            );
```

Comment:
When can this happen? Can there be a race and gas keys being removed after we unlock?

Also - if we are worries about the race, should we extract nonces in the loop above instead? When iterating using `get_raw_prefix_for_gas_keys` wouldn't we get all the gas keys which we could remove here instead of having to additionally get gas key here (by parsing out for public key and nonce)?

Reply:
So assuming that by different remove functions you mean remove_gas_key_nonce and remove_gas_key.

Both of them have similar enough implementations. If in remove_gas_key_nonce we change index type to Option<NonceIndex> then remove_gas_key would be the same as calling remove_gas_key_nonce with None index. Both gas keys with nonces and keys without nonces share the same trie key.

In fact parse_nonce_index_from_gas_key_key function already supports returning None (for non-nonce gas key). Note that we already know the size of the public key since we have to parse it out already so there is no need for trial and error.